inspirehep/modules/references/processors.py

```diff
@@ -0,0 +1,176 @@
+# along with INSPIRE. If not, see <http://www.gnu.org/licenses/>.
+#
+# In applying this licence, CERN does not waive the privileges and immunities
+# granted to it by virtue of its status as an Intergovernmental Organization
+# or submit itself to any jurisdiction.
+
+"""References Processors."""
+
+import re
+
+import idutils
+from isbn import ISBNError
+
+from inspirehep.utils.pubnote import split_pubnote
+
+
+def _is_arxiv(obj):
+    arxiv_test = obj.split()
+    if not arxiv_test:
+        return False
+    return idutils.is_arxiv(arxiv_test[0])
+
+
+def _normalize_arxiv(obj):
+    return idutils.normalize_arxiv(obj.split()[0])
+
+
+class ReferenceBuilder(object):
+    """Class used for building JSON reference objects given simple properties.
+
+    Use this when:
+        * Converting from MARC to Literature
+        * Parsing refextract output
+        * Pushing a record from Holdingpen
+
+    We wrote this in a non-pythonic non-generic way so it's extensible to any
+    format a reference field might take.
+    """
+
+    RE_VALID_CNUM = re.compile(r'C\d{2}-\d{2}-\d{2}(\.\d+)?')
+    RE_VALID_PUBNOTE = re.compile(r'.*,.*,.*(,.*)?')
+
+    def __init__(self):
+        self.obj = {}
+
+    def _ensure_field(self, field_name, value):
+        if field_name not in self.obj:
+            self.obj[field_name] = value
+
+    def set_number(self, number):
+        try:
+            self.obj['number'] = int(number)
+        except (ValueError, TypeError):
+            pass
+
+    def set_record(self, record):
+        self.obj['record'] = record
+        self._ensure_field('curated_relation', False)
+
+    def curate(self):
+        self.obj['curated_relation'] = True
+
+    def set_texkey(self, texkey):
+        self.obj['texkey'] = texkey
+
+    def add_title(self, title):
+        self._ensure_field('titles', [])
+        self.obj['titles'].append({'value': title})
+
+    def add_misc(self, misc):
+        self._ensure_field('misc', [])
+        self.obj['misc'].append(misc)
+
+    def add_raw_reference(self, raw_reference, serialization='text'):
+        self._ensure_field('raw_reference', [])
+        self.obj['raw_reference'].append({
+            'value': raw_reference,
+            'serialization': serialization})
+
+    def set_year(self, year):
+        try:
+            year = int(year)
+        except (ValueError, TypeError):
+            return
+        if year >= 1000 and year <= 2050:
+            self._ensure_field('publication_info', {})
+            self.obj['publication_info']['year'] = year
+
+    def add_url(self, url):
+        self._ensure_field('urls', [])
+        self.obj['urls'].append({'value': url})
+
+    def add_author(self, full_name, role=None):
+        self._ensure_field('authors', [])
+        author = {} if not role else {'role': role}
+        author['full_name'] = full_name
+        self.obj['authors'].append(author)
+
+    def set_pubnote(self, pubnote):
```

Comment:
I think that for the moment we might create pubnotes so we parse them back. But, the PRs for references will grow a lot if we affect the `hepcrawl -> holdingpen -> inspirehep` stack since there's a need for:
* unifying hepcrawl output
* changing holdingpen `publication` schema
* fixing holdingpen views
* fixing workflows to save the correct thing in our db

That can be a full and different PR